## Classifier To Detect Fake News

In [3]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.datasets import fetch_20newsgroups
from sklearn.model_selection import cross_val_score
from re import sub
from nltk.stem import PorterStemmer

from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [4]:
df = pd.read_csv("fake_or_real_news.csv")
df.head()

id                                              title  \
0   8476                       You Can Smell Hillary’s Fear   
1  10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2   3608        Kerry to go to Paris in gesture of sympathy   
3  10142  Bernie supporters on Twitter erupt in anger ag...   
4    875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [5]:
print("******** setup **********")
stemmer = PorterStemmer()

# extract into dataframes
text = pd.DataFrame(df.text, columns=['text'])
label = df['label']

******** setup **********


In [6]:
label.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [15]:
text.iloc[3,0]

"— Kaydee King (@KaydeeKing) November 9, 2016 The lesson from tonight's Dem losses: Time for Democrats to start listening to the voters. Stop running the same establishment candidates. \n— People For Bernie (@People4Bernie) November 9, 2016 If Dems didn't want a tight race they shouldn't have worked against Bernie. \n— Walker Bragman (@WalkerBragman) November 9, 2016 \nNew York Times columnist Paul Krugman, who was one of Hillary Clinton’s most outspoken surrogates during the contentious Democratic primary, blamed Clinton’s poor performance on Green Party candidate Jill Stein, who has so far received a negligible number of votes nationally, saying Stein was the Ralph Nader of 2016 in preventing a Clinton victory. The account @BerniesTeachers threw Krugman’s analysis back in his face. Your candidate was the issue. Take responsibility. https://t.co/KHyOuUSrFS \n— Teachers for Bernie (@BerniesTeachers) November 9, 2016 \nAna Navarro, a Republican who recently endorsed Hillary Clinton, sum

In [10]:
distances = euclidean_distances(docarray)
distances_df = pd.DataFrame(data=distances)
distances_df

0          1          2          3          4          5     \
0      0.000000  21.748563  22.045408  21.283797  20.736441  28.740216   
1     21.748563   0.000000  17.406895  17.146428  16.031220  27.037012   
2     22.045408  17.406895   0.000000  17.291616  15.427249  26.683328   
3     21.283797  17.146428  17.291616   0.000000  14.035669  27.147744   
4     20.736441  16.031220  15.427249  14.035669   0.000000  26.305893   
5     28.740216  27.037012  26.683328  27.147744  26.305893   0.000000   
6     21.977261  17.888544  17.291616  17.435596  15.779734  26.739484   
7     20.493902  15.198684  14.282857  14.317821  12.409674  26.038433   
8     26.115130  24.020824  23.237900  23.515952  22.759613  30.000000   
9     22.847319  18.894444  17.549929  18.734994  17.146428  27.748874   
10    21.954498  18.083141  17.720045  17.635192  15.937377  27.202941   
11    21.656408  17.088007  16.522712  16.000000  14.035669  26.962938   
12    21.189620  16.370706  15.588457  15.231546  13.964240  26.739484   
13    19.899749  14.317821  13.416408  13.527749  11.224972  25.651511   
14    19.773720  14.071247  13.076697  13.190906  10.908712  25.670995   
15    29.715316  28.035692  27.874720  27.640550  26.888659  32.511536   
16    20.396078  15.198684  14.696938  14.662878  12.489996  26.191602   
17    20.469489  15.556349  15.588457  15.033296  13.379088  26.057628   
18    23.345235  20.346990  19.672316  19.646883  18.466185  28.195744   
19    20.445048  15.394804  14.696938  14.247807  12.409674  26.000000   
20    23.790755  19.924859  19.493589  19.874607  18.973666  27.604347   
21    20.976177  15.779734  15.165751  14.866069  13.416408  26.457513   
22    22.781571  19.026298  18.303005  18.220867  16.703293  27.073973   
23    25.039968  22.583180  22.293497  22.271057  20.904545  29.240383   
24    23.958297  20.904545  20.297783  20.518285  19.026298  28.035692   
25    24.535688  21.377558  21.213203  21.000000  19.493589  28.913665   
26    20.566964  16.309506  15.588457  14.899664  12.767145  26.210685   
27    20.591260  15.842980  15.491933  15.394804  13.928388  26.305893   
28    22.271057  18.947295  18.439089  18.027756  16.852300  27.531800   
29    23.916521  19.924859  19.849433  19.313208  18.384776  27.928480   
...         ...        ...        ...        ...        ...        ...   
6305  20.639767  15.652476  14.560220  14.866069  12.727922  26.191602   
6306  20.469489  15.231546  14.456832  14.282857  12.369317  26.134269   
6307  22.090722  17.464249  16.124515  16.763055  15.362291  27.092434   
6308  21.931712  17.262677  16.822604  16.613248  14.798649  26.962938   
6309  30.692019  29.849623  28.913665  29.274562  28.844410  32.954514   
6310  23.811762  20.688161  19.723083  20.396078  19.261360  27.730849   
6311  25.632011  23.021729  22.248595  22.934690  22.113344  29.614186   
6312  24.248711  21.470911  20.736441  20.273135  19.442222  28.600699   
6313  22.781571  19.390719  18.520259  18.110770  16.462078  27.730849   
6314  20.880613  15.779734  15.684387  14.866069  13.266499  26.305893   
6315  24.576411  21.563859  20.445048  21.189620  20.297783  28.565714   
6316  23.515952  20.493902  20.223748  19.595918  18.520259  28.442925   
6317  20.024984  14.764823  13.747727  13.711309  11.704700  25.787594   
6318  19.949937  15.198684  14.899664  14.177447  12.727922  26.115130   
6319  21.679483  17.860571  17.888544  16.941074  15.165751  27.092434   
6320  22.022716  18.110770  17.521415  17.088007  16.031220  27.147744   
6321  28.792360  26.153394  25.238859  25.337719  24.637370  32.695565   
6322  20.273135  15.165751  14.247807  14.560220  12.688578  25.942244   
6323  24.248711  20.808652  20.000000  19.924859  18.708287  28.106939   
6324  21.142375  16.673332  15.968719  15.811388  14.247807  26.851443   
6325  21.771541  18.947295  18.867962  18.303005  17.549929  27.748874   
6326  20.518285  16.733201  16.881943  16.248077  15.459625  27.221315   
6327  21.587033

In [12]:
distances[:10, :10]

array([[ 0.        , 21.74856317, 22.04540769, 21.28379665, 20.73644135,
        28.74021573, 21.97726098, 20.49390153, 26.11512971, 22.84731932],
       [21.74856317,  0.        , 17.40689519, 17.1464282 , 16.03121954,
        27.03701167, 17.88854382, 15.19868415, 24.0208243 , 18.89444363],
       [22.04540769, 17.40689519,  0.        , 17.29161647, 15.42724862,
        26.68332813, 17.29161647, 14.28285686, 23.23790008, 17.54992877],
       [21.28379665, 17.1464282 , 17.29161647,  0.        , 14.03566885,
        27.14774392, 17.43559577, 14.31782106, 23.51595203, 18.734994  ],
       [20.73644135, 16.03121954, 15.42724862, 14.03566885,  0.        ,
        26.30589288, 15.77973384, 12.40967365, 22.75961335, 17.1464282 ],
       [28.74021573, 27.03701167, 26.68332813, 27.14774392, 26.30589288,
         0.        , 26.73948391, 26.03843313, 30.        , 27.74887385],
       [21.97726098, 17.88854382, 17.29161647, 17.43559577, 15.77973384,
        26.73948391,  0.        , 14.52583905

## KNN

In [16]:
print("******** prepare data **********")
new_data = []
for i in range(text.shape[0]):
    new_data.append(sub("[^a-zA-Z]", " ", text.iloc[i,0]))

lowercase_data = []
for i in range(len(new_data)):
    lowercase_data.append(new_data[i].lower())

stemmed_data = []
for i in range(len(lowercase_data)):
    words = lowercase_data[i].split()
    stemmed_words = []
    for w in words:
        stemmed_words.append(stemmer.stem(w))
    stemmed_data.append(" ".join(stemmed_words))

print("******** setup vector model **********")
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(analyzer = "word", binary = True, min_df=2, stop_words='english')
docarray = vectorizer.fit_transform(stemmed_data).toarray()

print("******** model and XV **********")
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors=4)

# do the 10-fold cross validation
scores = cross_val_score(model, docarray, label, cv=10)
print("Fold Accuracies: {}".format(scores))
print("XV Accuracy: {: 6.2f}%".format(scores.mean()*100))

******** prepare data **********
******** setup vector model **********
******** model and XV **********
Fold Accuracies: [0.68346457 0.6466877  0.65299685 0.670347   0.68246445 0.69352291
 0.65086888 0.70300158 0.65876777 0.67298578]
XV Accuracy:  67.15%


## Naive Bayes

In [9]:
print("******** prepare data **********")
new_data = []
for i in range(text.shape[0]):
    new_data.append(sub("[^a-zA-Z]", " ", text.iloc[i,0]))

lowercase_data = []
for i in range(len(new_data)):
    lowercase_data.append(new_data[i].lower())

stemmed_data = []
for i in range(len(lowercase_data)):
    words = lowercase_data[i].split()
    stemmed_words = []
    for w in words:
        stemmed_words.append(stemmer.stem(w))
    stemmed_data.append(" ".join(stemmed_words))

print("******** setup vector model **********")
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(analyzer = "word", binary = True, min_df=2, stop_words='english')
docarray = vectorizer.fit_transform(stemmed_data).toarray()
coords = vectorizer.get_feature_names()
docterm = pd.DataFrame(data=docarray,columns=coords)

print("******** model and XV **********")
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()


# do the 10-fold cross validation
scores = cross_val_score(model, docarray, label, cv=10)
print("Fold Accuracies: {}".format(scores))
print("XV Accuracy: {: 6.2f}%".format(scores.mean()*100))

******** prepare data **********
******** setup vector model **********
******** model and XV **********
Fold Accuracies: [0.89606299 0.89905363 0.88328076 0.87697161 0.88151659 0.87519747
 0.87677725 0.89099526 0.89099526 0.87993681]
XV Accuracy:  88.51%


### Analysis

There is a difference in accuracy of the models, the Naive Bayes model is the one that has higher accuracy. It has 88.51% of the accuracy but the KNN model only has 67.15% of the accuracy. The naive bayes is a conditional probability model and a generative classifier which thus gives a better statistical result than the KNN Model. Overall the Naive Bayes Model generally gives a faster time result than the KNN Model.